packages/user/TokenStream/service/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+            Decimal::new(self.total_deposited, self.precision())
+        }
+
+        pub async fn claimed(&self) -> Decimal {
+            Decimal::new(self.total_claimed, self.precision())
+        }
+
+        pub async fn vested(&self) -> Decimal {
+            Decimal::new(self.total_vested(), self.precision())
+        }
+
+        pub async fn claimable(&self) -> Decimal {
+            Decimal::new(self.balance_claimable(), self.precision())
+        }
+
+        pub async fn unclaimed(&self) -> Decimal {
+            Decimal::new(self.unclaimed_total(), self.precision())
+        }
+    }
+
+    impl Stream {
+        fn new(half_life_seconds: u32, token_id: u32) -> Self {
+            check(half_life_seconds > 0, "half life must be over 0");
+            let now = TransactSvc::call().currentBlock().time.seconds();
+
+            Self {
+                nft_id: Nfts::call().mint(),
+                token_id,
+                half_life_seconds,
+                total_deposited: 0.into(),
+                total_claimed: 0.into(),
+                last_deposit_timestamp: now,
+                claimable_at_last_deposit: 0.into(),
+            }
+        }
+
+        fn precision(&self) -> Precision {
+            psibase::services::tokens::Wrapper::call()
+                .getToken(self.token_id)
+                .precision
+                .try_into()
+                .unwrap()
+        }
+
+        pub fn add(decay_rate_per_million: u32, token_id: u32) -> Self {
+            let mut instance = Self::new(decay_rate_per_million, token_id);
+            instance.save();
+            instance
+        }
+
+        pub fn get(nft_id: u32) -> Option<Self> {
+            StreamTable::new().get_index_pk().get(&nft_id)
+        }
+
+        pub fn get_assert(nft_id: u32) -> Self {
+            check_some(Self::get(nft_id), "Stream does not exist")
+        }
+
+        pub fn check_is_owner(&self, account: AccountNumber) {
+            check(
+                psibase::services::nft::Wrapper::call()
+                    .getNft(self.nft_id)
+                    .owner
+                    == account,
+                format!("{} must be holder of NFT ID: {}", account, self.nft_id).as_str(),
+            );
+        }
+
+        fn decay_rate(&self) -> f64 {
+            std::f64::consts::LN_2 / self.half_life_seconds as f64
+        }
+
+        /// Total available to be claimed now (vested - already claimed)
+        pub fn balance_claimable(&self) -> Quantity {
+            self.total_vested()
+                .value
+                .saturating_sub(self.total_claimed.value)
+                .into()
+        }
+
+        /// Deposit `amount` into the bucket.
+        pub fn deposit(&mut self, amount: Quantity) {
+            self.claimable_at_last_deposit = self.total_vested();
+            self.total_deposited = self
+                .total_deposited
+                .value
+                .saturating_add(amount.value)
```

Comment:
If this actually saturates, only the amount that is actually added should be debited from the shared balance. Another option is to decrement `total_deposited` by `total_claimed`, and reset `total_claimed` to zero, which makes overflow impossible. Note that overflow is only possible by cycling the same tokens through the stream repeatedly.